swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift

```diff
@@ -0,0 +1,163 @@
+
+// --- stubs ---
+
+struct URL
+{
+	init?(string: String) {}
+	init?(string: String, relativeTo: URL?) {}
+}
+
+extension String {
+	init(contentsOf: URL) throws {
+        var data = ""
+
+        // ...
+
+        self.init(data)
+    }
+}
+
+var SQLITE_OK : Int32 = 0
+var SQLITE_TRANSIENT : (@convention(c) (UnsafeMutableRawPointer?) -> Void)?
+
+func sqlite3_exec(
+		_ _: OpaquePointer?,
+		_ sql: UnsafePointer<CChar>?,
+		_ callback: (@convention(c) (UnsafeMutableRawPointer?, Int32, UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?, UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?) -> Int32)?,
+		_ _: UnsafeMutableRawPointer?,
+		_ errmsg: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafePointer<CChar>?,
+		_ nByte: Int32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafePointer<CChar>?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare_v2(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafePointer<CChar>?,
+		_ nByte: Int32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafePointer<CChar>?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare_v3(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafePointer<CChar>?,
+		_ nByte: Int32,
+		_ prepFlags: UInt32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafePointer<CChar>?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare16(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafeRawPointer?,
+		_ nByte: Int32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafeRawPointer?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare16_v2(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafeRawPointer?,
+		_ nByte: Int32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafeRawPointer?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare16_v3(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafeRawPointer?,
+		_ nByte: Int32,
+		_ prepFlags: UInt32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafeRawPointer?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_bind_text(
+		_ _: OpaquePointer?,
+		_ _: Int32,
+		_ _: UnsafePointer<CChar>?,
+		_ _: Int32,
+		_ callback: (@convention(c) (UnsafeMutableRawPointer?) -> Void)?
+	) -> Int32 { return SQLITE_OK }
+// (many other `sqlite3_bind_*` functions exist)
+
+func sqlite3_step(
+		_ _: OpaquePointer?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_finalize(
+		_ pStmt: OpaquePointer?
+	) -> Int32 { return SQLITE_OK }
+
+func sanitize(_ string: String) -> String { return string }
```

Comment:
I'm not sure how `sanitize` should work, though. Is it a test-only function to test sanitizers in general, or do we expect the query to recognize any function called `sanitize` or similar as a valid sanitizer?

Reply:
The query is likely to recognize functions using a heuristic such as the function name, so matching functions called `sanitize`, `filter` etc (`escape`, perhaps).  The tricky part is that's likely to include some bad sanitizers (like just adding quotes to the ends of the string, or escaping some but not all important characters) that we would rather flag somehow.

Reply:
I see that in Java we have a different approach - `PrimitiveType`, `BoxedType`, `NumberType` are sanitizers.  I assume this is because they can't possibly contain or expand to free text as would be required for an injection attack.

Reply:
I'm unclear whether it's *possible* to write a safe sanitizer for SQL, or whether in fact the prepared statement approach is the only 'right' way?

Reply:
You _technically_ could, but it's not recommended because it's easy to do it wrong. In general, the suggested approach always is using prepared statements (or wrapper libraries that do not handle raw SQL) when possible, or otherwise using a well-established security library to do the sanitization/validation for you.